engine/lib/phases/phase_drop_return_break_continue.ml

```diff
@@ -0,0 +1,164 @@
+
+    let visitor =
+      let module Visitors = Ast_visitors.Make (F) in
+      object (self)
+        inherit [_] Visitors.map as _super
+
+        method! visit_expr (in_loop : (loop_info * A.ty) option) e =
+          let span = e.span in
+          match (e.e, in_loop) with
+          | Return { e; _ }, None -> e
+          (* we know [e] is on an exit position: the return is
+             thus useless, we can skip it *)
+          | Let { monadic = None; lhs; rhs; body }, _ ->
+              let body = self#visit_expr in_loop body in
+              {
+                e with
+                e = Let { monadic = None; lhs; rhs; body };
+                typ = body.typ;
+              }
+              (* If a let expression is an exit node, then it's body
+                 is as well *)
+          | Match { scrutinee; arms }, _ ->
+              let arms = List.map ~f:(self#visit_arm in_loop) arms in
+              let typ =
+                match arms with { arm; _ } :: _ -> arm.body.typ | [] -> e.typ
+              in
+              { e with e = Match { scrutinee; arms }; typ }
+          | If { cond; then_; else_ }, _ ->
+              let then_ = self#visit_expr in_loop then_ in
+              let else_ = Option.map ~f:(self#visit_expr in_loop) else_ in
+              { e with e = If { cond; then_; else_ }; typ = then_.typ }
+          | Return { e; _ }, Some ({ return_type; break_type }, acc_type) ->
+              UA.M.expr_Constructor_CF ~return_type ~span ~break_type ~e
+                ~acc:{ e with typ = acc_type } `Return
+          | ( Break { e; acc = Some (_, acc); _ },
+              Some ({ return_type; break_type }, _) ) ->
+              UA.M.expr_Constructor_CF ~return_type ~span ~break_type ~e ~acc
+                `Break
+          | ( Continue { acc = Some (_, acc); _ },
+              Some ({ return_type; break_type }, _) ) ->
+              UA.M.expr_Constructor_CF ~return_type ~span ~break_type ~acc
+                `Continue
+          | _, Some ({ return_type; break_type }, _)
+            when Option.is_some return_type || Option.is_some break_type ->
+              UA.M.expr_Constructor_CF ~return_type ~span ~break_type ~acc:e
+                `Continue
+          | _ -> e
+        (** The invariant here is that [visit_expr] is called only
+                 on expressions that are on exit positions. [visit_expr]
+                 is first called on root expressions, which are (by
+                 definition) exit nodes. Then, [visit_expr] itself makes
+                 recursive calls to sub expressions that are themselves
+                 in exit nodes. **)
+      end
+
+    [%%inline_defs dmutability + dsafety_kind]
+
+    let rec dexpr' (span : span) (expr : A.expr') : B.expr' =
+      match expr with
+      | [%inline_arms "dexpr'.*" - Return - Break - Continue - Loop] -> auto
+      | Return _ | Break _ | Continue _ ->
+          Error.raise { kind = NonTrivialAndMutFnInput (*TODO Correct*); span }
```

Comment:
This TODO is about changing the error?
Here you should use `Error.assertion_failure span "Return/Break/Continue are expected to be gone as this point"` or something like that